[SPARK-10515] When killing executor, the pending replacement executors should not be lost

KaiXinXiaoLei opened the pull request:
If the heartbeat receiver kills executors (and new ones are not registered to replace them), the idle timeout for the old executors will be lost (and then change a total number of executors requested by Driver), So new ones will be not to asked to replace them.
For example, executorsPendingToRemove=Set(1), and executor 2 is idle timeout before a new executor is asked to replace executor 1. Then driver kill executor 2, and sending RequestExecutors to AM. But executorsPendingToRemove=Set(1,2), So AM doesn't allocate a executor to replace 1.

see: https://github.com/apache/spark/pull/8668

Commit:
add unit test

Commit:
add numPendingExecutors

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43124 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43124/consoleFull) for   PR 8945 at commit [`1bdde8e`](https://github.com/apache/spark/commit/1bdde8e88ed36194e020e773db415e65181357a9).

SparkQA commented:
[Test build #43124 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43124/console) for   PR 8945 at commit [`1bdde8e`](https://github.com/apache/spark/commit/1bdde8e88ed36194e020e773db415e65181357a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43124/
Test FAILed.

vanzin commented:
Hmm, test failure looks like it might be related.

Also, does this replace #8668? If so, could you close one of them?

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -369,6 +369,35 @@ class StandaloneDynamicAllocationSuite
     assert(apps.head.getExecutorLimit === 1)
   }
 
+test("the pending replacement executors should not be lost (SPARK-10515)") {
```
indentation is off

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -369,6 +369,35 @@ class StandaloneDynamicAllocationSuite
+  }
+  // sync executors between the Master and the driver, needed because
+  // the driver refuses to kill executors it does not know about
+  syncExecutors(sc)
```
indentation is off

Commit:
indentation  style, thanks

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43389 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43389/consoleFull) for   PR 8945 at commit [`e382315`](https://github.com/apache/spark/commit/e382315e9905b735837bc37a63acd16b72242ada).

SparkQA commented:
[Test build #43389 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43389/console) for   PR 8945 at commit [`e382315`](https://github.com/apache/spark/commit/e382315e9905b735837bc37a63acd16b72242ada).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43389/
Test FAILed.

KaiXinXiaoLei commented:
jenkins test please.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43453/consoleFull) for   PR 8945 at commit [`e382315`](https://github.com/apache/spark/commit/e382315e9905b735837bc37a63acd16b72242ada).

andrewor14 commented:
Thanks LGTM. I'll merge this once tests pass.

vanzin commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -369,6 +369,35 @@ class StandaloneDynamicAllocationSuite
+    assert(master.apps.head.executors.size === 2)
+
+    assert(sc.killExecutor(executors.head))
+    assert(master.apps.head.executors.size === 2)
```
This has the same problem I mentioned before in the other PR. `sc.killExecutor` does not immediately update the state of the `master.apps` list. So these tests are bound to fail in weird ways.

You need to use `killNExecutors` instead of `sc.killExecutor` and `getApplications` instead of `master.apps`. See other tests in this same file.

SparkQA commented:
[Test build #43453 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43453/console) for   PR 8945 at commit [`e382315`](https://github.com/apache/spark/commit/e382315e9905b735837bc37a63acd16b72242ada).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43453/
Test PASSed.

Commit:
change test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43458/consoleFull) for   PR 8945 at commit [`b7b42cc`](https://github.com/apache/spark/commit/b7b42ccbd8364fedbd652e6763fd3c566539bf93).

SparkQA commented:
[Test build #43458 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43458/console) for   PR 8945 at commit [`b7b42cc`](https://github.com/apache/spark/commit/b7b42ccbd8364fedbd652e6763fd3c566539bf93).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43458/
Test FAILed.

Commit:
change test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43464/consoleFull) for   PR 8945 at commit [`cb69dc5`](https://github.com/apache/spark/commit/cb69dc5b1795e49a1f10db2db0e7fe8a5ba6e5cc).

SparkQA commented:
[Test build #43464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43464/console) for   PR 8945 at commit [`cb69dc5`](https://github.com/apache/spark/commit/cb69dc5b1795e49a1f10db2db0e7fe8a5ba6e5cc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43464/
Test PASSed.

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -369,6 +369,38 @@ class StandaloneDynamicAllocationSuite
+    // kill executor 1
+    assert(sc.killExecutor(executors.head))
+    apps = getApplications()
+    assert(apps.head.executors.size === 2)
```
hm, not sure if I understand this. Why does the number of executors stay at 2 after you call `sc.killExecutor`, which does not replace it? Shouldn't it go down to 1?

KaiXinXiaoLei replied:
@andrewor14 Here i use sc.killExecutor(executors.head), I want to say a executor lost and a new executor should start to replace. Before the new executor registers, the executor is idle timeout. Then the total number of executors should not change. So  "apps.head.executors.size === 2"

vanzin commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -369,6 +369,38 @@ class StandaloneDynamicAllocationSuite
+    // kill executor 1, and replace it
+    assert(sc.killAndReplaceExecutor(executors.head))
+    var apps = getApplications()
+    assert(apps.head.executors.size === 2)
```
I think this code is racy; the executors count goes down to 1 before coming back to 2, and depending on the order of the events happening in the background, this test might fail. Using `eventually` here might be safer, since you do want the count to eventually come back to 2.

KaiXinXiaoLei replied:
@vanzin 
```
   assert(sc.killAndReplaceExecutor(executors.head))
    var apps = getApplications()
    assert(apps.head.executors.size === 2)
```
I want  to say, kill a executor and a new executor should replaces it. The total number of executor should be not changed.

vanzin replied:
I understand that. But what I'm saying is, in the master's view of the world, what happens is that one executor goes away and a new executor comes up. So there may be a window in which the executor count, as seen from the master, is "1", temporarily, before the new executor starts. And that would cause your test to fail sporadically.

KaiXinXiaoLei replied:
@vanzin I understand. Can you tell me how to change? Thanks.

vanzin replied:
Just follow examples in the same file. e.g.:

    eventually(timeout(10.seconds), interval(10.millis)) {
      val apps = getApplications()
      assert(apps.size === 1)
      assert(apps.head.id === appId)
      assert(apps.head.executors.size === 2)
      assert(apps.head.getExecutorLimit === Int.MaxValue)
    }

KaiXinXiaoLei replied:
@vanzin I have changed. Please review. Thanks.

Commit:
Make sure the  total number not changed.

Kill a executor and a new executor should replaces it.  Make sure the  total number of executor be not changed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43628/consoleFull) for   PR 8945 at commit [`da13040`](https://github.com/apache/spark/commit/da13040aca20f0c739be8958675f5b39dac4d82c).

SparkQA commented:
[Test build #43628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43628/console) for   PR 8945 at commit [`da13040`](https://github.com/apache/spark/commit/da13040aca20f0c739be8958675f5b39dac4d82c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43628/
Test PASSed.

vanzin commented:
LGTM, will let Andrew have a final look.

andrewor14 commented:
Forgot to add: merged into master 1.5.
